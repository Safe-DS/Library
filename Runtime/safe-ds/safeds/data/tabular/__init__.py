from ._column import Column, ColumnStatistics
from ._row import Row
from ._table import Table
