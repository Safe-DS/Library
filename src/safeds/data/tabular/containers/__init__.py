from ._column import Column
from ._row import Row
from ._table import Table
from ._tagged_table import TaggedTable
