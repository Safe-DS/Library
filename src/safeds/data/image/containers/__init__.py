"""Classes that can store image data."""

from ._image import Image

__all__ = [
    "Image",
]
