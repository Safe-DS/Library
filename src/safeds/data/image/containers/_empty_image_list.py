from __future__ import annotations

import warnings
from typing import TYPE_CHECKING, Self

from safeds._utils import _structural_hash
from safeds.data.image.containers._image_list import ImageList
from safeds.data.image.containers._single_size_image_list import _SingleSizeImageList
from safeds.data.image.utils._image_transformation_error_and_warning_checks import (
    _check_add_noise_errors,
    _check_adjust_brightness_errors_and_warnings,
    _check_adjust_color_balance_errors_and_warnings,
    _check_adjust_contrast_errors_and_warnings,
    _check_blur_errors_and_warnings,
    _check_crop_errors_and_warnings,
    _check_remove_images_with_size_errors,
    _check_resize_errors,
    _check_sharpen_errors_and_warnings,
)
from safeds.exceptions import IndexOutOfBoundsError

if TYPE_CHECKING:
    from pathlib import Path

    from torch import Tensor

    from safeds.data.image.containers import Image


class _EmptyImageList(ImageList):
    """
    An ImageList is a list of different images. It can hold different sizes of Images. The channel of all images is the same.

    This is the class for an empty ImageList.

    To create an `ImageList` call one of the following static methods:

    | Method                                                                        | Description                                              |
    | ----------------------------------------------------------------------------- | -------------------------------------------------------- |
    | [from_images][safeds.data.image.containers._image_list.ImageList.from_images] | Create an ImageList from a list of Images.               |
    | [from_files][safeds.data.image.containers._image_list.ImageList.from_files]   | Create an ImageList from a directory or a list of files. |
    """

    _instance = None

    @staticmethod
    def _warn_empty_image_list() -> None:
        """Warn if a transform method is used on an empty ImageList."""
        warnings.warn(
            "You are using an empty ImageList. This method changes nothing if used on an empty ImageList.",
            UserWarning,
            stacklevel=2,
        )

    def __new__(cls) -> Self:
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    @staticmethod
    def _create_image_list(images: list[Tensor], indices: list[int]) -> ImageList:
        raise NotImplementedError

    def _clone(self) -> ImageList:
        return _EmptyImageList()

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, ImageList):
            return NotImplemented
        return isinstance(other, _EmptyImageList)

    def __hash__(self) -> int:
        return _structural_hash("_EmptyImageList")

    def __sizeof__(self) -> int:
        return 0

    @property
    def number_of_images(self) -> int:
        return 0

    @property
    def widths(self) -> list[int]:
        return []

    @property
    def heights(self) -> list[int]:
        return []

    @property
    def channel(self) -> int:
        return NotImplemented

    @property
    def number_of_sizes(self) -> int:
        return 0

    def get_image(self, index: int) -> Image:
        raise IndexOutOfBoundsError(index)

    def index(self, _image: Image) -> list[int]:
        return []

    def has_image(self, _image: Image) -> bool:
        return False

    def to_jpeg_files(self, _path: str | Path | list[str | Path]) -> None:
        warnings.warn("You are using an empty ImageList. No files will be saved.", UserWarning, stacklevel=2)

    def to_png_files(self, _path: str | Path | list[str | Path]) -> None:
        warnings.warn("You are using an empty ImageList. No files will be saved.", UserWarning, stacklevel=2)

    def to_images(self, _indices: list[int] | None = None) -> list[Image]:
        return []

    def change_channel(self, channel: int) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        if channel not in (1, 3, 4):
            raise ValueError(f"Channel {channel} is not a valid channel option. Use either 1, 3 or 4")
        return _EmptyImageList()

    def _add_image_tensor(self, image_tensor: Tensor, _index: int) -> ImageList:
        return _SingleSizeImageList._create_image_list([image_tensor], [0])

    def add_images(self, images: list[Image] | ImageList) -> ImageList:
        return ImageList.from_images(images) if isinstance(images, list) else images

    def remove_image_by_index(self, index: int | list[int]) -> ImageList:
        raise IndexOutOfBoundsError(index)

    def _remove_image_by_index_ignore_invalid(self, _index: int | list[int]) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        return _EmptyImageList()

    def remove_images_with_size(self, width: int, height: int) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        _check_remove_images_with_size_errors(width, height)
        return _EmptyImageList()

    def remove_duplicate_images(self) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        return _EmptyImageList()

    def shuffle_images(self) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        return _EmptyImageList()

    def resize(self, new_width: int, new_height: int) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        _check_resize_errors(new_width, new_height)
        return _EmptyImageList()

    def convert_to_grayscale(self) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        return _EmptyImageList()

    def crop(self, x: int, y: int, width: int, height: int) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        _check_crop_errors_and_warnings(
            x,
            y,
            width,
            height,
            x + 1,
            y + 1,
            plural=True,
        )  # Disable x|y >= min_width|min_height check with min_width|min_height=x|y+1
        return _EmptyImageList()

    def flip_vertically(self) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        return _EmptyImageList()

    def flip_horizontally(self) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        return _EmptyImageList()

    def adjust_brightness(self, factor: float) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        _check_adjust_brightness_errors_and_warnings(factor, plural=True)
        return _EmptyImageList()

    def add_noise(self, standard_deviation: float) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        _check_add_noise_errors(standard_deviation)
        return _EmptyImageList()

    def adjust_contrast(self, factor: float) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        _check_adjust_contrast_errors_and_warnings(factor, plural=True)
        return _EmptyImageList()

    def adjust_color_balance(self, factor: float) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        _check_adjust_color_balance_errors_and_warnings(factor, 0, plural=True)  # Disable channel check with channel=0
        return _EmptyImageList()

    def blur(self, radius: int) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        _check_blur_errors_and_warnings(radius, 1, plural=True)
        return _EmptyImageList()

    def sharpen(self, factor: float) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        _check_sharpen_errors_and_warnings(factor, plural=True)
        return _EmptyImageList()

    def invert_colors(self) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        return _EmptyImageList()

    def rotate_right(self) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        return _EmptyImageList()

    def rotate_left(self) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        return _EmptyImageList()

    def find_edges(self) -> ImageList:
        _EmptyImageList._warn_empty_image_list()
        return _EmptyImageList()
