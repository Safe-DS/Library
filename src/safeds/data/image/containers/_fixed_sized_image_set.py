from __future__ import annotations

import copy
from pathlib import Path
from typing import TYPE_CHECKING

import torch
from torch import Tensor
from torchvision.transforms.v2 import functional as func2

from safeds.data.image.containers import Image
from safeds.data.image.containers._image_set import ImageSet
from safeds.exceptions import DuplicateIndexError, IndexOutOfBoundsError

if TYPE_CHECKING:
    from safeds.data.image.containers._various_sized_image_set import _VariousSizedImageSet


class _FixedSizedImageSet(ImageSet):

    def __init__(self):
        self._tensor = None
        self._tensor_positions_to_indices: list[int] = []  # list[tensor_position] = index
        self._indices_to_tensor_positions: dict[int, int] = {}  # {index: tensor_position}

    @staticmethod
    def _create_image_set(images: list[Tensor], indices: list[int]) -> ImageSet:
        image_set = _FixedSizedImageSet()
        images_ready_to_concat = []
        images_with_correct_channel = []
        images_with_less_channels = []
        max_channel = 0
        for image, index in zip(images, indices):
            if max_channel < image.size(dim=-3):  # all images have to increase their channel
                images_with_less_channels += images_with_correct_channel
                images_with_correct_channel = [(image, index)]
                max_channel = image.size(dim=-3)
            elif max_channel > image.size(dim=-3):  # current image has to increase its channel
                images_with_less_channels.append((image, index))
            else:  # current image has same channel as max_channel
                images_with_correct_channel.append((image, index))
        for image, index in images_with_correct_channel:
            images_ready_to_concat.append((image.unsqueeze(dim=0), index))
        for image, index in images_with_less_channels:
            if max_channel == 3:  # image channel 1 and max channel 3
                image = torch.cat([image, image, image], dim=0)
            elif image.size(dim=0) == 1:  # image channel 1 and max channel 4
                image = torch.cat([image, image, image, torch.full(image.size(), 255)], dim=0)
            else:  # image channel 3 and max channel 4
                image = torch.cat([image, torch.full(image[0:1].size(), 255)], dim=0)
            images_ready_to_concat.append((image.unsqueeze(dim=0), index))
        image_set._tensor = torch.cat([image for image, index in images_ready_to_concat])
        image_set._tensor_positions_to_indices = [index for image, index in images_ready_to_concat]
        image_set._indices_to_tensor_positions = image_set._calc_new_indices_to_tensor_positions()

        return image_set

    def clone(self) -> ImageSet:
        cloned_image_set = self._clone_without_tensor()
        cloned_image_set._tensor = self._tensor.detach().clone()
        return cloned_image_set

    def _clone_without_tensor(self) -> _FixedSizedImageSet:
        cloned_image_set = _FixedSizedImageSet()
        cloned_image_set._indices_to_tensor_positions = copy.deepcopy(self._indices_to_tensor_positions)
        cloned_image_set._tensor_positions_to_indices = copy.deepcopy(self._tensor_positions_to_indices)
        return cloned_image_set

    def _calc_new_indices_to_tensor_positions(self):
        _indices_to_tensor_positions: dict[int, int] = {}
        for i, index in enumerate(self._tensor_positions_to_indices):
            _indices_to_tensor_positions[index] = i
        return _indices_to_tensor_positions

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, ImageSet):
            return NotImplemented
        if not isinstance(other, _FixedSizedImageSet):
            return False
        return (
            self._tensor.size() == other._tensor.size()
            and set(self._tensor_positions_to_indices) == set(self._tensor_positions_to_indices)
            and set(self._indices_to_tensor_positions) == set(self._indices_to_tensor_positions)
            and torch.all(torch.eq(self._tensor[torch.tensor(self._tensor_positions_to_indices).argsort()], other._tensor[torch.tensor(other._tensor_positions_to_indices).argsort()])).item()
        )

    @property
    def number_of_images(self) -> int:
        return self._tensor.size(dim=0)

    @property
    def widths(self) -> list[int]:
        return [self._tensor.size(dim=3)]

    @property
    def heights(self) -> list[int]:
        return [self._tensor.size(dim=2)]

    @property
    def channel(self) -> int:
        return self._tensor.size(dim=1)

    @property
    def number_of_sizes(self) -> int:
        return 1

    def get_image(self, index: int) -> Image:
        if index not in self._indices_to_tensor_positions:
            raise IndexOutOfBoundsError(index)
        return Image(self._tensor[self._indices_to_tensor_positions[index]])

    def index(self, image: Image) -> list[int]:
        if image not in self:
            return []
        return [self._tensor_positions_to_indices[i] for i in (image._image_tensor == self._tensor).all(dim=[1, 2, 3]).nonzero()]

    def has_image(self, image: Image) -> bool:
        return not (image.width != self.widths[0] or image.height != self.heights[0]) and image._image_tensor in self._tensor

    def to_jpeg_files(self, path: str | Path | list[str] | list[Path]) -> None:
        pass

    def to_png_files(self, path: str | Path | list[str] | list[Path]) -> None:
        pass

    def to_images(self, indices: list[int] | None = None) -> list[Image]:
        if indices is None:
            indices = self._tensor_positions_to_indices
        else:
            wrong_indices = []
            for index in indices:
                if index not in self._indices_to_tensor_positions:
                    wrong_indices.append(index)
            if len(wrong_indices) == 0:
                raise IndexOutOfBoundsError(wrong_indices)
        return [Image(self._tensor[self._indices_to_tensor_positions[index]]) for index in indices]

    def change_channel(self, channel: int) -> ImageSet:
        image_set = self._clone_without_tensor()
        if self.channel == channel:
            image_set._tensor = self._tensor
        elif self.channel == 1 and channel == 3:
            image_set._tensor = torch.cat([self._tensor, self._tensor, self._tensor], dim=1)
        elif self.channel == 1 and channel == 4:
            image_set._tensor = torch.cat([self._tensor, self._tensor, self._tensor, torch.full(self._tensor.size(), 255)], dim=1)
        elif self.channel in (3, 4) and channel == 1:
            image_set._tensor = self.convert_to_grayscale()._as_fixed_sized_image_set()._tensor[:, 0:1]
        elif self.channel == 3 and channel == 4:
            image_set._tensor = torch.cat([self._tensor, torch.full(self._tensor[:, 0:1].size(), 255)], dim=1)
        elif self.channel == 4 and channel == 3:
            image_set._tensor = self._tensor[:, 0:3]
        else:
            image_set._tensor = self._tensor
        return image_set

    def _add_image_tensor(self, image_tensor: Tensor, index: int) -> ImageSet:
        from safeds.data.image.containers import _VariousSizedImageSet

        if index in self._indices_to_tensor_positions:
            raise DuplicateIndexError(index)

        if self._tensor.size(dim=2) == image_tensor.size(dim=1) and self._tensor.size(dim=3) == image_tensor.size(dim=2):
            image_set = self._clone_without_tensor()
            if image_tensor.size(dim=0) > self.channel:
                tensor = self.change_channel(image_tensor.size(dim=0))._as_fixed_sized_image_set()._tensor
            else:
                tensor = self._tensor
            if image_tensor.size(dim=0) < tensor.size(dim=1):
                if tensor.size(dim=1) == 3:
                    image_set._tensor = torch.cat([tensor, torch.cat([image_tensor, image_tensor, image_tensor], dim=0).unsqueeze(dim=0)])
                if tensor.size(dim=0) == 1:
                    image_set._tensor = torch.cat([tensor, torch.cat([image_tensor, image_tensor, image_tensor, torch.full(image_tensor.size(), 255)], dim=0).unsqueeze(dim=0)])
                else:
                    image_set._tensor = torch.cat([tensor, torch.cat([image_tensor, torch.full(image_tensor[0:1].size(), 255)], dim=0).unsqueeze(dim=0)])
            else:
                image_set._tensor = torch.cat([tensor, image_tensor.unsqueeze(dim=0)])
            image_set._tensor_positions_to_indices.append(index)
            image_set._indices_to_tensor_positions[index] = len(self)
        else:
            image_set = _VariousSizedImageSet()
            image_set._image_set_dict[(self.widths[0], self.heights[0])] = self.clone()
            image_set._image_set_dict[(image_tensor.size(dim=2), image_tensor.size(dim=1))] = _FixedSizedImageSet._create_image_set([image_tensor], [index])

            if image_tensor.size(dim=0) != self.channel:
                image_set = image_set.change_channel(max(image_tensor.size(dim=0), self.channel))
            for index in self._tensor_positions_to_indices:
                image_set._indices_to_image_size_dict[index] = (self.widths[0], self.heights[0])
            image_set._indices_to_image_size_dict[index] = (image_tensor.size(dim=2), image_tensor.size(dim=1))

        return image_set

    def add_images(self, images: list[Image] | ImageSet) -> ImageSet:
        max_index = max(self._tensor_positions_to_indices)
        new_indices = list(range(max_index, max_index + 1 + len(images)))
        if isinstance(images, list):
            images = ImageSet.from_images(images, new_indices)
        if isinstance(images, _VariousSizedImageSet):
            image_set = _VariousSizedImageSet()
            image_set._image_set_dict[(self.widths[0], self.heights[0])] = self.clone()
            for index in self._tensor_positions_to_indices:
                image_set._indices_to_image_size_dict[index] = (self.widths[0], self.heights[0])
            return image_set.add_images(images)
        if isinstance(images, _FixedSizedImageSet):
            if images.widths == self.widths and images.heights == self.heights:
                return _FixedSizedImageSet._create_image_set([image._image_tensor for image in self.to_images() + images.to_images()], self._tensor_positions_to_indices + new_indices)
            else:
                image_set = _VariousSizedImageSet()
                image_set._image_set_dict[(self.widths[0], self.heights[0])] = self.clone()
                image_set._image_set_dict[(images.widths[0], images.heights[0])] = images.clone()
                for index in self._tensor_positions_to_indices:
                    image_set._indices_to_image_size_dict[index] = (self.widths[0], self.heights[0])
                for index in images._tensor_positions_to_indices:
                    image_set._indices_to_image_size_dict[index] = (images.widths[0], images.heights[0])

                if images.channel != self.channel:
                    image_set = image_set.change_channel(max(images.channel, self.channel))
                return image_set

    def remove_image_by_index(self, index: int | list[int]) -> ImageSet:
        if isinstance(index, int):
            index = [index]
        image_set = _FixedSizedImageSet()
        image_set._tensor = self._tensor[[i for i, v in enumerate(self._tensor_positions_to_indices) if v not in index]].clone()
        image_set._tensor_positions_to_indices = [i - len([k for k in index if k < i]) for i in self._tensor_positions_to_indices if i not in index]  # TODO: should the indices be changed?
        image_set._indices_to_tensor_positions = image_set._calc_new_indices_to_tensor_positions()
        return image_set

    def remove_images_with_size(self, width: int, height: int) -> ImageSet:
        if self.widths[0] == width and self.heights[0] == height:
            return _FixedSizedImageSet()
        else:
            return self.clone()

    def remove_duplicate_images(self) -> ImageSet:
        image_set = _FixedSizedImageSet()
        image_set._tensor, new_indices = self._tensor.unique(dim=0, return_inverse=True)
        offset = 0
        indices = []
        for index_n, index_v in enumerate(new_indices):
            if list(new_indices).index(index_v) == index_n:
                indices.append(self._tensor_positions_to_indices[index_v + offset] - offset)
            else:
                offset += 1
        image_set._tensor_positions_to_indices = indices
        image_set._indices_to_tensor_positions = image_set._calc_new_indices_to_tensor_positions()
        return image_set

    def shuffle_images(self) -> ImageSet:
        pass

    def resize(self, new_width: int, new_height: int) -> ImageSet:
        pass

    def convert_to_grayscale(self) -> ImageSet:
        image_set = self._clone_without_tensor()
        image_set._tensor = func2.rgb_to_grayscale(self._tensor[:, 0:3])
        return image_set

    def crop(self, x: int, y: int, width: int, height: int) -> ImageSet:
        pass

    def flip_vertically(self) -> ImageSet:
        pass

    def flip_horizontally(self) -> ImageSet:
        pass

    def adjust_brightness(self, factor: float) -> ImageSet:
        pass

    def add_noise(self, standard_deviation: float) -> ImageSet:
        pass

    def adjust_contrast(self, factor: float) -> ImageSet:
        pass

    def adjust_color_balance(self, factor: float) -> ImageSet:
        pass

    def blur(self, radius: int) -> ImageSet:
        pass

    def sharpen(self, factor: float) -> ImageSet:
        pass

    def invert_colors(self) -> ImageSet:
        pass

    def rotate_right(self) -> ImageSet:
        pass

    def rotate_left(self) -> ImageSet:
        pass

    def find_edges(self) -> ImageSet:
        pass
