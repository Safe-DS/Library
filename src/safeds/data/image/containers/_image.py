from __future__ import annotations

import io
from pathlib import Path
from typing import BinaryIO

from PIL.Image import Image as PillowImage
from PIL.Image import open as open_image

from safeds.data.image.typing import ImageFormat


class Image:
    """
    A container for image data.

    Parameters
    ----------
    data : BinaryIO
        The image data as bytes.
    """

    @staticmethod
    def from_jpeg_file(path: str) -> Image:
        """
        Create an image from a JPEG file.

        Parameters
        ----------
        path : str
            The path to the JPEG file.

        Returns
        -------
        image : Image
            The image.
        """
        return Image(
            data=Path(path).open("rb"),
            format_=ImageFormat.JPEG,
        )

    @staticmethod
    def from_png_file(path: str) -> Image:
        """
        Create an image from a PNG file.

        Parameters
        ----------
        path : str
            The path to the PNG file.

        Returns
        -------
        image : Image
            The image.
        """
        return Image(
            data=Path(path).open("rb"),
            format_=ImageFormat.PNG,
        )

    def __init__(self, data: BinaryIO, format_: ImageFormat):
        data.seek(0)

        self._image: PillowImage = open_image(data, formats=[format_.value])
        self._format: ImageFormat = format_

    # ------------------------------------------------------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------------------------------------------------------

    @property
    def format(self) -> ImageFormat:
        """
        Get the image format.

        Returns
        -------
        format : ImageFormat
            The image format.
        """
        return self._format

    # ------------------------------------------------------------------------------------------------------------------
    # Conversion
    # ------------------------------------------------------------------------------------------------------------------

    def to_jpeg_file(self, path: str) -> None:
        """
        Save the image as a JPEG file.

        Parameters
        ----------
        path : str
            The path to the JPEG file.
        """
        Path(path).parent.mkdir(parents=True, exist_ok=True)
        self._image.save(path, format="jpeg")

    def to_png_file(self, path: str) -> None:
        """
        Save the image as a PNG file.

        Parameters
        ----------
        path : str
            The path to the PNG file.
        """
        Path(path).parent.mkdir(parents=True, exist_ok=True)
        self._image.save(path, format="png")

    # ------------------------------------------------------------------------------------------------------------------
    # IPython integration
    # ------------------------------------------------------------------------------------------------------------------

    def _repr_jpeg_(self) -> bytes | None:
        """
        Return a JPEG image as bytes.

        If the image is not a JPEG, return None.

        Returns
        -------
        jpeg : bytes
            The image as JPEG.
        """
        if self._format != ImageFormat.JPEG:
            return None

        buffer = io.BytesIO()
        self._image.save(buffer, format="jpeg")
        buffer.seek(0)
        return buffer.read()

    def _repr_png_(self) -> bytes | None:
        """
        Return a PNG image as bytes.

        If the image is not a PNG, return None.

        Returns
        -------
        png : bytes
            The image as PNG.
        """
        if self._format != ImageFormat.PNG:
            return None

        buffer = io.BytesIO()
        self._image.save(buffer, format="png")
        buffer.seek(0)
        return buffer.read()
