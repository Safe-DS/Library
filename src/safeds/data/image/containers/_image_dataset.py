from __future__ import annotations

import copy

import numpy as np
import torch
from torch import Tensor

from safeds._config import _get_device
from safeds.data.image.containers import ImageList
from safeds.data.image.containers._single_size_image_list import _SingleSizeImageList
from safeds.data.tabular.containers import Table
from safeds.exceptions import NonNumericColumnError, OutputLengthMismatchError, IndexOutOfBoundsError


class ImageDataset:

    def __init__(self, input_data: ImageList, output_data: ImageList | Table, batch_size=1, shuffle=False) -> None:
        self._shuffle_tensor_indices = torch.LongTensor(list(range(len(input_data))))
        self._shuffle_after_epoch = shuffle
        self._batch_size = batch_size
        self._next_batch_index = 0

        if not isinstance(input_data, _SingleSizeImageList):
            raise ValueError("The given input ImageList contains images of different sizes.")
        else:
            self._input = input_data
        if (isinstance(output_data, Table) and len(input_data) != output_data.number_of_rows) or (isinstance(output_data, ImageList) and len(input_data) != len(output_data)):
            raise OutputLengthMismatchError(f"{len(input_data)} != {output_data.number_of_rows if isinstance(output_data, Table) else len(output_data)}")
        if isinstance(output_data, Table):
            non_numerical_columns = []
            wrong_interval_columns = []
            for column_name in output_data.column_names:
                if not output_data.get_column_type(column_name).is_numeric():
                    non_numerical_columns.append(column_name)
                elif output_data.get_column(column_name).minimum() < 0 or output_data.get_column(column_name).maximum() > 1:
                    wrong_interval_columns.append(column_name)
            if len(non_numerical_columns) > 0:
                raise NonNumericColumnError(f"Columns {non_numerical_columns} are not numerical.")
            if len(wrong_interval_columns) > 0:
                raise ValueError(f"Columns {wrong_interval_columns} have values outside of the interval [0, 1].")
            _output = _TableAsTensor(output_data)
        elif isinstance(output_data, _SingleSizeImageList):
            _output = output_data.clone()._as_single_size_image_list()
        else:
            raise ValueError("The given output ImageList contains images of different sizes.")
        self._output = _output

    def _get_batch(self, batch_number: int, batch_size: int | None = None) -> tuple[Tensor, Tensor]:
        if batch_size is None:
            batch_size = self._batch_size
        if batch_size * batch_number >= len(self._input):
            raise IndexOutOfBoundsError(batch_size * batch_number)
        max_index = batch_size * (batch_number + 1) if batch_size * (batch_number + 1) < len(self._input) else len(self._input)
        input_tensor = self._input._tensor[self._shuffle_tensor_indices[[self._input._indices_to_tensor_positions[index] for index in range(batch_size * batch_number, max_index)]]].to(torch.float32) / 255
        output_tensor: Tensor
        if isinstance(self._output, _SingleSizeImageList):
            output_tensor = self._output._tensor[self._shuffle_tensor_indices[[self._output._indices_to_tensor_positions[index] for index in range(batch_size * batch_number, max_index)]]].to(torch.float32) / 255
        else:  # _output is instance of _TableAsTensor
            output_tensor = self._output._tensor[self._shuffle_tensor_indices[batch_size * batch_number:max_index]]
        return input_tensor, output_tensor

    def __iter__(self) -> ImageDataset:
        # self._batch_index = 0
        # if self._shuffle_after_epoch:
        #     self._shuffle_inplace()
        # return self

        # def _generator():
        #     batch_index = 0
        #
        #     while batch_index * self._batch_size < len(self._input):
        #         yield self._get_batch(batch_index)
        #
        #         batch_index += 1
        if self._shuffle_after_epoch:
            im_ds = self.shuffle()
        else:
            im_ds = copy.copy(self)
        im_ds._next_batch_index = 0
        return im_ds

    def __next__(self) -> tuple[Tensor, Tensor]:
        if self._next_batch_index * self._batch_size >= len(self._input):
            raise StopIteration
        self._next_batch_index += 1
        return self._get_batch(self._next_batch_index - 1)

    def __len__(self) -> int:
        return self._input.number_of_images

    def shuffle(self) -> ImageDataset:
        im_dataset: ImageDataset = copy.copy(self)
        im_dataset._shuffle_tensor_indices = torch.randperm(len(self))
        im_dataset._next_batch_index = 0
        return im_dataset

    # def _shuffle_inplace(self) -> None:
    #     self._shuffle_tensor_indices = torch.randperm(len(self))
    #
    # def _reset_indices_inplace(self) -> None:
    #     self._shuffle_tensor_indices = torch.LongTensor(list(range(len(self))))


class _TableAsTensor:

    def __init__(self, table: Table) -> None:
        self._column_names = table.column_names

        columns_as_tensors = []
        for column_name in table.column_names:
            columns_as_tensors.append(torch.Tensor(table.get_column(column_name)._data.values.astype(np.float32)).unsqueeze(dim=0))

        self._tensor = torch.cat(columns_as_tensors, dim=0).to(_get_device()).T

        if not torch.all(self._tensor.sum(dim=1) == torch.ones(self._tensor.size(dim=0))):
            raise ValueError("The given table is not correctly one hot encoded as it contains rows that have a sum not equal to 1.")
