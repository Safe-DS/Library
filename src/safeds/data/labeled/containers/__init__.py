"""Classes that can store labeled data."""

from typing import TYPE_CHECKING

import apipkg

if TYPE_CHECKING:
    from ._dataset import Dataset
    from ._image_dataset import ImageDataset
    from ._tabular_dataset import TabularDataset
    from ._time_series_dataset import TimeSeriesDataset

apipkg.initpkg(
    __name__,
    {
        "Dataset": "._dataset:Dataset",
        "ImageDataset": "._image_dataset:ImageDataset",
        "TabularDataset": "._tabular_dataset:TabularDataset",
        "TimeSeriesDataset": "._time_series_dataset:TimeSeriesDataset",
    },
)

__all__ = [
    "Dataset",
    "ImageDataset",
    "TabularDataset",
    "TimeSeriesDataset",
]
