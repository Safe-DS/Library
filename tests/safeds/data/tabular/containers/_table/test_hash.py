import pytest
from safeds.data.tabular.containers import Table


@pytest.mark.parametrize(
    ("table1", "table2"),
    [
        (Table(), Table()),
        (Table({"a": [], "b": []}), Table({"a": [], "b": []})),
        (Table({"col1": [1]}), Table({"col1": [1]})),
        (Table({"col1": [1, 2, 3]}), Table({"col1": [1, 1, 3]})),
    ],
    ids=[
        "empty table",
        "rowless table",
        "equal tables",
        "different values",
    ],
)
def test_should_return_same_hash_for_equal_tables(table1: Table, table2: Table) -> None:
    assert hash(table1) == hash(table2)


@pytest.mark.parametrize(
    ("table1", "table2"),
    [
        (Table({"col1": [1]}), Table({"col2": [1]})),
        (Table({"col1": [1, 2, 3]}), Table({"col1": ["1", "2", "3"]})),
        (Table({"col1": [1, 2, 3]}), Table({"col1": [1, 2, 3, 4]})),
    ],
    ids=["different column names", "different types", "different number of rows"],
)
def test_should_return_different_hash_for_unequal_tables(table1: Table, table2: Table) -> None:
    assert hash(table1) != hash(table2)
