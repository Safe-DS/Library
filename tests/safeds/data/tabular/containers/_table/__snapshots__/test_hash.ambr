# serializer version: 1
# name: TestSameHash.test_should_return_same_hash_in_different_processes[empty]
  1789859531466043636
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[no rows]
  585695607399955642
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[with data]
  909875695937937648
# ---
